[jit] Support nn.GRU in script

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23643 [jit] make nn.LSTM accept PackedSequence instead of Tuples
* **#23266 [jit] Support nn.GRU in script**

Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

wanchaol committed:
[jit] Support nn.GRU in script

zou3519 commented on torch/nn/modules/rnn.py:
```diff
@@ -655,10 +657,68 @@ class GRU(RNNBase):
+                             self.num_layers, self.dropout, self.training, self.bidirectional)
+        output = result[0]
+
+        output = result[0]
```
copy/paste of above line?

zou3519 commented on torch/nn/modules/rnn.py:
```diff
@@ -655,10 +657,68 @@ class GRU(RNNBase):
 
+    def forward_impl(self, input, hx, batch_sizes, max_batch_size, sorted_indices):
+        # type: (Tensor, Optional[Tensor], Optional[Tensor], int, Optional[Tensor]) -> Tuple[Tensor, Tensor]  # noqa
+        if hx is None:
```
It's not so nice that all of these have to be added in, but I'm sure there are reasons

wanchaol replied:
yeah that's kind of unfortunate, I originally want to directly script the RNNBase instead of GRU separately, but the dict to dispatch different module is a dict(str, function), JIT dictionary does not support function/module as a value now so we could only do it separately..

wanchaol committed:
Update on "[jit] Support nn.GRU in script"

wanchaol committed:
Update on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

wanchaol committed:
Update on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

wanchaol committed:
Update on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

wanchaol committed:
refactor on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

driazati commented on test/test_jit.py:
```diff
@@ -12539,6 +12539,24 @@ def forward(self, input):
+                self.x = torch.nn.GRU(5, 5)
+
+            @torch.jit.script_method
+            def forward(self, input: PackedSequence) -> Tuple[PackedSequence, torch.Tensor]:
```
This needs to use Python 2 syntax

driazati commented on torch/nn/modules/rnn.py:
```diff
@@ -655,10 +657,66 @@ class GRU(RNNBase):
+
+        self.check_forward_args(input, hx, batch_sizes)
+        if batch_sizes is None:
+            result = _VF.gru(input, hx, self._get_flat_weights(), self.bias, self.num_layers,
```
We could do it where there's something like `RNNBase.run_impl` where in `RNNBase` it's something like
```python
def run_impl(self, input_batch, sizes...):
	impl = self.impls(self.mode)
	return impl(...)
```

and override that in `GRU` to directly call `_VF.gru`, the JIT should only see the overridden version

```python
def run_impl(self, ...):
	return _VF.gru(...)
```

If this works it'd cut down the duplicated code a lot

driazati commented on test/test_jit.py:
```diff
@@ -12539,6 +12539,24 @@ def forward(self, input):
+            def forward(self, input: PackedSequence) -> Tuple[PackedSequence, torch.Tensor]:
+                return self.x(input)
+
+        eager_out = self.runAndSaveRNG(lambda x: torch.nn.GRU(5, 5)(x), (input,))[0]
```
Can you also test the non-PackedSequence version?

driazati reviewed (approved):
The code duplication isn't great but I don't think we have a way around it. Looks good, I'd say give the overriding the implementation (comment below) a shot but if that doesn't work then this is fine as is

wanchaol committed:
Update on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

wanchaol committed:
rebase with master on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

wanchaol committed:
address comments on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

wanchaol committed:
Update on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

eellison commented on test/test_jit.py:
```diff
@@ -12803,6 +12803,40 @@ def forward(self, input):
+        tensor_script_out = self.runAndSaveRNG(lambda x: TensorGRU()(x), (tensor_input,))[0]
+
+        self.assertEqual(seq_eager_out, seq_script_out)
+        self.assertEqual(tensor_eager_out, tensor_script_out)
```
Can you add tests for import and export ? Or just use `checkScript`

wanchaol replied:
I think we have setup module/function hooks for JIT test cases for import and export, so we don't need to add it here. This test follows the same syntax with `test_nn_LSTM` above, in order to use `checkScript`, we will need to refactor it to be able to freeze the rng state inside, which `checkScript` does not do it currently. I will see if I could make it in a follow up PR.

wanchaol committed:
Update on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

apaszke reviewed (changes requested):
Eh, can't we add union types to script? Honestly this is a _significant_ regression in the code quality. I understand that it's an important use case for the JIT, but this is just unreasonable.

wanchaol commented:
> Eh, can't we add union types to script? Honestly this is a _significant_ regression in the code quality. I understand that it's an important use case for the JIT, but this is just unreasonable.

@apaszke nn.LSTM (subclass of RNNBase) already use this forward dispatch mechanism on master now.. I am just moving this to `RNNBase` directly instead of `LSTM` (I planned to remove that code in a follow up PR), which is a bad code quality there already. I agree that we need to add Union type, but that's something need careful consideration as it changes bunch of things like type refinements, etc.

apaszke commented:
I'm not saying it doesn't require careful consideration, but this is basically destroying the codebase. The fact that we've already regressed is not an argument supporting further regressions. If you see something can't be expressed but is important then make it expressible instead of hacking your way through the path of least resistance. I can't imagine explaining what the hell is going on there to potential contributors...

eellison commented:
I agree this isn't very readable. Can you add comments to the places you think someone wouldn't be able to understand without any context on jit ? 

After you add the comments I think we should land. It is important that we are able to script all of our nn Modules & it is confusing to have scriptable LSTMs but not GRUs (especially after advertising on our [blog](https://pytorch.org/blog/optimizing-cuda-rnn-with-torchscript/)). union types would be a significant drain on developer productivity & time, potentially confusing to users (ex. invariant lists), and currently impossible without other significant changes (like removing list specializations). We haven't had any issues asking for them by users either. 

I think in follow up PRs we should look into what we can do on the JIT side to make this easier to express & more readable.

zou3519 commented:
> After you add the comments I think we should land. It is important that we are able to script all of our nn Modules & it is confusing to have scriptable LSTMs but not GRUs (especially after advertising on our [blog](https://pytorch.org/blog/optimizing-cuda-rnn-with-torchscript/)).

I want to point out that what we advertised on the blog was optimizing custom RNNs, not the nn Module versions of RNNs.

eellison commented:
Another possibility is to only change GRU and not RNNBase (or maybe add shared class that LSTM & GRU inherit from)

wanchaol commented:
> I'm not saying it doesn't require careful consideration, but this is basically destroying the codebase. The fact that we've already regressed is not an argument supporting further regressions. If you see something can't be expressed but is important then make it expressible instead of hacking your way through the path of least resistance. I can't imagine explaining what the hell is going on there to potential contributors...

@apaszke I agree that this code something not quite readable compared to the code before, I could think of how to support Union type in script, not sure how significance the change would be. 

I did this PR first on the GRU class directly without modifying the RNNBase, see [here](https://github.com/pytorch/pytorch/pull/23266/commits/d46146cc44af131bae11a3d3c18481dcdb94b7a4), then I realized there's too much code duplication which not good either, so I changed the RNNBase directly. If you think that was a good approach or a shared class that LSTM & GRU inherit from as @eellison suggested, I can make the change.

apaszke commented:
Having the change be limited to a single class would be an improvement. It should also be accompanied by a big comment describing why this looks like that, and saying it's a transition state only (e.g. we wanted to make it on time for the release). Finally, the statement in the comment saying that it's only temporary should be true, meaning that a fix for this should be on a roadmap.

apaszke commented:
cc @zdevito

wanchaol committed:
make change limit to GRU class and add comment for temporary state on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

wanchaol committed:
Update on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)

eellison commented on torch/nn/modules/rnn.py:
```diff
@@ -369,6 +370,16 @@ def __init__(self, *args, **kwargs):
+# 3. This is temporary only and in the transition state that we want to make it
+#    on time for the release
+#
+# TODO: remove the overriding implementations for LSTM and GRU when TorchScript
```
Could you remove the mention of Unions in this section? That is not the only way we could make this more expressible, and we haven't as a team made any promises about supporting it. Same thing with Any Type.

wanchaol committed:
Update on "[jit] Support nn.GRU in script"


Differential Revision: [D16466586](https://our.internmc.facebook.com/intern/diff/D16466586)
